chain/default_syncer.go

```diff
@@ -302,42 +302,51 @@ func (syncer *DefaultSyncer) widen(ctx context.Context, ts types.TipSet) (types.
 	// Lookup tipsets with the same parents from the store.
 	parentSet, err := ts.Parents()
 	if err != nil {
-		return nil, err
+		return types.NoTipSet, err
 	}
 	height, err := ts.Height()
 	if err != nil {
-		return nil, err
+		return types.NoTipSet, err
 	}
 	if !syncer.chainStore.HasTipSetAndStatesWithParentsAndHeight(parentSet.String(), height) {
-		return nil, nil
+		return types.NoTipSet, nil
 	}
 	candidates, err := syncer.chainStore.GetTipSetAndStatesByParentsAndHeight(parentSet.String(), height)
 	if err != nil {
-		return nil, err
+		return types.NoTipSet, err
 	}
 	if len(candidates) == 0 {
-		return nil, nil
+		return types.NoTipSet, nil
 	}
 
 	// Only take the tipset with the most blocks (this is EC specific logic)
-	max := candidates[0]
+	max := candidates[0].TipSet
 	for _, candidate := range candidates[0:] {
-		if len(candidate.TipSet) > len(max.TipSet) {
-			max = candidate
+		if candidate.TipSet.Len() > max.Len() {
+			max = candidate.TipSet
 		}
 	}
 
-	// Add blocks of the biggest tipset in the store to a copy of ts
-	wts := ts.Clone()
-	for _, blk := range max.TipSet {
-		if err = wts.AddBlock(blk); err != nil {
-			return nil, err
-		}
+	// Form a new tipset from the union of ts and the largest in the store, de-duped.
+	blocks := make(map[cid.Cid]*types.Block)
+	for i := 0; i < ts.Len(); i++ {
+		blocks[ts.At(i).Cid()] = ts.At(i)
+	}
+	for i := 0; i < max.Len(); i++ {
+		blocks[max.At(i).Cid()] = max.At(i)
+	}
+	var blockSlice []*types.Block
+	for _, b := range blocks {
+		blockSlice = append(blockSlice, b)
+	}
```

Comment:
I'd find this logic a little more straightforward if the first loop added to both the slice and the map, and the second loop added to the slice only if the cid wasn't in the map (and there was no 3rd loop). It's debatable.

Changing it would be more efficient, but that's not worth much here.